lib/cisco_node_utils/client/ssh/client.rb

```diff
@@ -0,0 +1,167 @@
+# SSH client library.
+#
+# November 2014, Glenn F. Matthews
+#
+# Copyright (c) 2014-2016 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require_relative '../client'
+require 'json'
+
+include Cisco::Logger
+
+# Class representing an SSH client connecting to a nx-os device.
+class Cisco::Client::SSH < Cisco::Client
+  register_client(self)
+
+  # Constructor for Client. This connects to the
+  # device via ssh
+  def initialize(**kwargs)
+    # rubocop:disable Style/HashSyntax
+    super(data_formats: [:nxapi_structured, :cli],
+          platform:     :nexus,
+          **kwargs)
+    @verify_host_key = (Gem.loaded_specs['net-ssh'].version < Gem::Version.create('4.2.0')) ? :paranoid : :verify_host_key
+    @host = kwargs[:host]
+    @username = kwargs[:username]
+    @password = kwargs[:password]
+    @port = kwargs[:port] || 22
+    @timeout = kwargs[:timeout] || 30
+  end
+
+  def self.validate_args(**kwargs)
+    super
+  end
+
+  # Clear the cache of CLI output results.
+  #
+  # If cache_auto is true (default) then this will be performed automatically
+  # whenever a set() is called, but providers may also call this
+  # to explicitly force the cache to be cleared.
+  def cache_flush
+    @cache_hash = {
+      'cli_conf'       => {},
+      'cli_show'       => {},
+      'cli_show_ascii' => {},
+    }
+  end
+
+  # Configure the given CLI command(s) on the device.
+  #
+  # @raise [RequestNotSupported] if this client doesn't support CLI config
+  #
+  # @param data_format one of Cisco::DATA_FORMATS. Default is :cli
+  # @param context [String, Array<String>] Zero or more configuration commands
+  #   used to enter the desired CLI sub-mode
+  # @param values [String, Array<String>] One or more commands
+  #   to enter within the CLI sub-mode.
+  # @param kwargs data-format-specific args
+  def set(data_format: :cli,
+          context: nil,
+          values: nil,
+          **_kwargs)
+    # we don't currently support nxapi_structured for configuration
+    fail Cisco::RequestNotSupported if data_format == :nxapi_structured
```

Comment:
I would prefer that we change this to  `structured_data` that can be used for both `ssh` and `nxapi` or any future clients.  I know this would result in more sweeping changes but we should probably by hiding the connection details from our `get_data_format` keys in the feature yaml files.

The `cach_flush` method also uses terms that are very nxapi specific so I think we should have a larger discussion around how we design for multiple clients.